Transceiver firmware upgrade testplan

az-pz opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->
Create CMIS CDB firmware upgrade testplan from the transceiver onboarding testplan and organize the two testplans in a separate directory. Stress testing of txvr firmware upgrade was also added.

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

az-pz committed:
Carve out txvr upgrade testplan.

az-pz committed:
Remove comments.

mssonicbld commented:
/azp run

az-pz committed:
Update link.

az-pz committed:
Move cli commands.

az-pz committed:
Add stress tests.

az-pz committed:
Remove firmware mgmt from onboarding testplan.

mssonicbld commented:
/azp run

az-pz committed:
Add test attributes.

az-pz committed:
Fix formatting.

prgeor commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,489 @@
+- CMIS
+- C-CMIS
+
+## Testbed Topology
```
@az-pz We don't need to duplicate this here since this is already specified in the transciever onboarding doc. you can reference that here.

prgeor reviewed (commented):
@az-pz are you taking care of the prerequisite where firmware download need to ensure the link is oper down before, during and after the CDB firmware download step?

prgeor commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,489 @@
+| restore_initial_firmwares | Bool | False | No | A flag indicating whether to restore the initial active and inactive firmware versions after testing is completed |
+| firmware_download_stress_iterations | Int | 5 | No | The number of iterations to stress test the firmware download process |
+| firmware_activation_stress_iterations | Int | 5 | No | The number of iterations to stress test the firmware activation process |
+| firmware_download_abort_method | String | "ctrl_c" | No | The method to abort the firmware download process. It can be one of the following strings : "ctrl_c", "sfputil_reset", "optic_reinsert" |
```
@az-pz keep default as `sfputil_reset`?

prgeor commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,489 @@
+| restore_initial_firmwares | Bool | False | No | A flag indicating whether to restore the initial active and inactive firmware versions after testing is completed |
+| firmware_download_stress_iterations | Int | 5 | No | The number of iterations to stress test the firmware download process |
+| firmware_activation_stress_iterations | Int | 5 | No | The number of iterations to stress test the firmware activation process |
+| firmware_download_abort_method | String | "ctrl_c" | No | The method to abort the firmware download process. It can be one of the following strings : "ctrl_c", "sfputil_reset", "optic_reinsert" |
```
@az-pz @mihirpat1 what about link up time timeout after new firmware activation?

az-pz replied:
We can reuse `port_wait_time_after_startup_sec` attribute from `Link related tests` attribute category. Updated the TC#2 with the attribute.

az-pz committed:
Resolve PR comments.

mssonicbld commented:
/azp run

az-pz committed:
Resolve PR comments.

az-pz commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -110,7 +109,7 @@ These tests do not require traffic and are standalone, designed to run on a Devi
     - `physical_port`: The physical port number on the DUT where the transceiver is connected (e.g., 1, 2, etc.).
     - `vendor_pn`: The vendor part number as specified in the transceiver's EEPROM.
-    - `normalized_vendor_pn`: The normalized vendor part number, created by applying the normalization rules described in the [CMIS CDB Firmware Binary Management](#141-cmis-cdb-firmware-binary-management) section.
+    - `normalized_vendor_pn`: The normalized vendor part number, created by applying the normalization rules described in the [CMIS CDB Firmware Binary Management](./cmis_cdb_firmware_upgrade_testplan.md#111-cmis-cdb-firmware-binary-management) section.
```
Link is working fine. No punctuation character except `-` is allowed in anchors. See https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#section-links .

az-pz committed:
Resolve PR comments.

mssonicbld commented:
/azp run

az-pz commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -170,50 +169,15 @@ These tests do not require traffic and are standalone, designed to run on a Devi
+    - `normalized_vendor_name`: The normalized vendor name, created by applying the normalization rules described in the [CMIS CDB Firmware Binary Management](./cmis_cdb_firmware_upgrade_testplan.md#111-cmis-cdb-firmware-binary-management) section.
     <br>The normalization rules ensure that the vendor name is consistent and compatible with directory structures used in firmware management and upgrade tests.
-    - `normalized_vendor_pn`: The normalized vendor part number, created by applying the normalization rules described in the [CMIS CDB Firmware Binary Management](#141-cmis-cdb-firmware-binary-management) section.
+    - `normalized_vendor_pn`: The normalized vendor part number, created by applying the normalization rules described in the [CMIS CDB Firmware Binary Management](./cmis_cdb_firmware_upgrade_testplan.md#111-cmis-cdb-firmware-binary-management) section.
```
See above comment : https://github.com/sonic-net/sonic-mgmt/pull/20102#discussion_r2270551413 .

az-pz committed:
Update test.

mssonicbld commented:
/azp run

az-pz commented:
> @az-pz are you taking care of the prerequisite where firmware download need to ensure the link is oper down before, during and after the CDB firmware download step?

@prgeor , The port should be oper up before the firmware download starts and it should stay oper up during and after the firmware download with no link flaps observed during the process.

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,445 @@
+
+| Attribute | Type | Default | Mandatory | Description |
+|-----------|------|---------|------------|-------------|
+| port_under_test | Dict | None | Yes | A dictionary containing the device name as the key and list of ports to be tested as its value |
```
@az-pz Can you please refer to https://github.com/sonic-net/sonic-mgmt/pull/20230 and update this table accordingly since we may not need a dictionary for storing many attribute in this table.

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,445 @@
+| Attribute | Type | Default | Mandatory | Description |
+|-----------|------|---------|------------|-------------|
+| port_under_test | Dict | None | Yes | A dictionary containing the device name as the key and list of ports to be tested as its value |
+| firmware_versions | Dict | None | Yes | A dictionary containing the normalized transceiver product number as the key and list of firmware versions to be tested as its value |
```
@az-pz Can you please add the below attributes
1. sleep_after_dom_disable_sec
2. monitor_kernel_errors
3. thermalctld_disabling_required
4. i2c_wait_time_post_reset

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,445 @@
+
+| TC No. | Test | Steps | Expected Results |
+|------|------|------|------------------|
+| 1 | Firmware download validation | 1. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>2. Wait until CLI execution completes | 1. CLI execution should finish within `transceiver_download_timeout_minutes` minutes and return 0 <br>2. Active FW version should remain unchanged<br>3. Inactive FW version should reflect the downloaded firmware version<br> 4. No link ...
```
"The kernel has no error messages in syslog" - this can be checked if monitor_kernel_errors is set.

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,445 @@
+| TC No. | Test | Steps | Expected Results |
+|------|------|------|------------------|
+| 1 | Firmware download validation | 1. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>2. Wait until CLI execution completes | 1. CLI execution should finish within `transceiver_download_timeout_minutes` minutes and return 0 <br>2. Active FW version should remain unchanged<br>3. Inactive FW version should reflect the downloaded firmware version<br> 4. No link ...
+| 2 | Firmware activation validation | 1. Shut down all the interfaces part of the physical ports<br>2. Execute firmware run<br>3. Execute firmware commit<br>4. Reset the transceiver and wait for 5 seconds<br>5. Startup all the interfaces in Step 1 | 1. The return code on step 2 and 3 is 0 (Return code 0 indicates success)<br>2. Active firmware version should now match the previous inactive firmw ...
```
@az-pz Can we verify the active and inactive FW image bank after each FW run and commit to ensure it is updated inline with expectation?

az-pz replied:
Split this test into two and added verifications steps for image banks.

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,445 @@
+|------|------|------|------------------|
+| 1 | Firmware download validation | 1. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>2. Wait until CLI execution completes | 1. CLI execution should finish within `transceiver_download_timeout_minutes` minutes and return 0 <br>2. Active FW version should remain unchanged<br>3. Inactive FW version should reflect the downloaded firmware version<br> 4. No link ...
+| 2 | Firmware activation validation | 1. Shut down all the interfaces part of the physical ports<br>2. Execute firmware run<br>3. Execute firmware commit<br>4. Reset the transceiver and wait for 5 seconds<br>5. Startup all the interfaces in Step 1 | 1. The return code on step 2 and 3 is 0 (Return code 0 indicates success)<br>2. Active firmware version should now match the previous inactive firmw ...
+|3 | Firmware download validation with invalid firmware binary | Download an invalid firmware binary (any file not released by the vendor) | 1. The active firmware version does not change<br>2. The inactive firmware version remains unchanged or is set to `0.0.0` or `N/A`<br> 3.  No change in "Committed Image"<br>4. No link flap should be seen<br>5. The kernel has no error messages in syslog<br>6. ...
```
@az-pz Can you elaborate on how do you plan to generate invalid FW binary?

az-pz commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,489 @@
+
+**Attributes for  the below tests**
+
+Following test attributes under `transceiver_firmware_upgrade` attribute category is applicable:
```
Yes, this is gonna be a json file.

mihirpat1 commented:
@az-pz Can you add a TC to issue CDB FW version read command 1000 times via API and ensure that the active and inactive FW versions are in line with the expectation?

az-pz committed:
Pull latest txvr onboarding testplan

az-pz commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,489 @@
+
+##### 1.1.1 CMIS CDB Firmware Binary Management
+
+###### 1.1.1.1 Firmware Binary Naming Guidelines
```
In our offline discussion, we decided that we'll be focusing only on dual bank scenario.

az-pz committed:
Minor changes.

az-pz commented:
This PR is dependent on #20230 . This PR should be merged only after #20230 .

mihirpat1 reviewed (commented):
@az-pz In the FW upgrade tests, can we also ensure to report the FW download, run and commit time for every test?
We can capture this requirement in the HLD

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,290 @@
+
+| TC No. | Test | Steps | Expected Results |
+|------|------|------|------------------|
+| 1 | Firmware download validation | 1. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>2. Wait until CLI execution completes | 1. CLI execution should finish within `firmware_download_timeout_minutes` minutes and return 0 <br>2. Active FW version should remain unchanged<br>3. Inactive FW version should reflect the downloaded firmware version<br> 4. No link fl ...
```
@az-pz Can we also ensure that the pid of `xcvcrd` remains unchanged as part of FW upgrade?

az-pz committed:
Address PR comments.

mssonicbld commented:
/azp run

az-pz commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,290 @@
+
+Following test attributes under `transceiver_firmware_upgrade` attribute category are applicable:
+
+| Attribute Name | Type | Default | Mandatory | Description |
```
Added. Thanks for the suggestion.

az-pz committed:
Add timeout for run and commit commands.

mssonicbld commented:
/azp run

az-pz committed:
Resolve PR comments.

az-pz committed:
Revise firmware binary storage structure in documentation

Updated the directory structure for CMIS CDB firmware binaries to include firmware version numbers.

az-pz committed:
Revise firmware upgrade test plan structure and URLs

Updated firmware directory structure and download URL format to include firmware version number.

mssonicbld commented:
/azp run

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,299 @@
+| firmware_download_abort_percentage | List | `[10, 50, 90]` | No | transceivers | The percentage of download progress at which the firmware download should be aborted. |
+| sleep_after_dom_disable_sec | Int | 5 | No | transceivers | The number of seconds to sleep after disabling DOM monitoring before proceeding with the test. |
+| monitor_kernel_errors | Bool | False | No | transceivers | A flag indicating whether to monitor kernel errors during the test. |
+| thermalctld_disabling_required | Bool | False | No | transceivers | A flag indicating whether to disable the thermalctld during the test. |
```
Why would this need to be disabled?

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,299 @@
+   | QSFP-100G-AOC-10YY          | QSFP-100G-AOC-GENERIC_2_ENDYY           | 10 has 2 digits, YY suffix preserved |
+   | QSFP-100G-AOC-100           | QSFP-100G-AOC-GENERIC_3_END             | 100 has 3 digits, no unit |
+   | QSFP-100G-AOC-3M            | QSFP-100G-AOC-GENERIC_1_ENDM            | 3 has 1 digit, M unit preserved |
+   | SFP-1000M                   | SFP-GENERIC_4_ENDM                      | 1000 has 4 digits, M unit preserved |
```
Why 1000M but 100G?

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,299 @@
+
+| Original Vendor Name    | Normalized Vendor Name | Explanation |
+|------------------------|------------------------|-------------|
+| ACME Corp.             | ACME_CORP              | Space and dot replaced with underscore |
```
. is not replaced by an underscore in this example. Perhaps should specify why in explanation.

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,299 @@
+**Prerequisites:**
+
+1. **DOM polling must be disabled** to prevent race conditions between I2C transactions and the CDB mode for modules that cannot support CDB background mode. Test should wait for `sleep_after_dom_disable_sec` seconds after disabling DOM to avoid the race condition.
+2. **Platform-specific processes:** On some platforms, `thermalctld` or similar user processes that perform I2C transactions with the module may need to be stopped if `thermalctld_disabling_required` flag is set.
```
Could we make this safe by detecting and stopping them automatically or something?

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -98,87 +97,400 @@ These tests do not require traffic and are standalone, designed to run on a Devi
+        "field_5": "value_5",
+        "field_6": "value_6"
+      },
+      "platform_specific": {
```
Is this one file for global community? Will this have many many platforms in it? Or is this just our injection point for local when we compute the anisble files?

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
Do we need two NICs for this topology? Could we do this with less equipment?

pfallon-nexthop commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -98,87 +97,400 @@ These tests do not require traffic and are standalone, designed to run on a Devi
-                "normalized_vendor_name": "normalized_vendor_name",
-                'vdm_supported': True,
-                'cdb_backgroundmode_supported': True,
-                'dual_bank_supported': True
```
Are we now making this a requirement and an assumption?

pfallon-nexthop reviewed (commented):
Thank you! We have a few comments and questions

az-pz committed:
Add test for cdb abort.

az-pz commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -225,7 +537,8 @@ The following tests aim to validate the link status and stability of transceiver
 | Issue CLI command to startup a port | Validate link status using CLI configuration | Ensure that the link is up and the port appears in the LLDP table. |
-| In a loop, issue startup/shutdown command 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
+| In a loop, issue startup/shutdown command for a port 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
+| In a loop, issue startup/shutdown command for all ports 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately for all relevant ports with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
```
This testcase was already present in the testplan. Once #20230 is implemented, this value can be controlled by a configurable attribute.

az-pz committed:
Update attribute.

az-pz committed:
Add test fw read in parallel.

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,302 @@
+
+| TC No. | Test | Steps | Expected Results |
+|------|------|------|------------------|
+| 1 | Firmware download validation | 1. Run CDB abort command if `firmware_download_cdb_abort_support` flag is set. | 2. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>3. Wait until CLI execution completes | 1. If CDB abort command was executed, it should be successful. <br>2. CLI execution should finish within `firmware_download_timeout_minutes` minutes and  ...
```
Shouldn't the below come under the steps section?

2. Download the next firmware specified in `firmware_versions` using the sfputil CLI<br>3. Wait until CLI execution completes

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,302 @@
+| 2 | Firmware run validation | 1. Shut down all the interfaces part of the physical ports<br>2. Execute firmware run<br><br>3. Reset the transceiver and wait for 5 seconds<br>4. Startup all the interfaces in Step 1 | 1. Firmware run command should finish in 60 seconds and the return code should be 0 (Return code 0 indicates success)<br>2. Active firmware version should now match the previous ina ...
+| 3 | Firmware commit validation | 1. Execute firmware commit for an interface| 1. Firmware commit command should finish in 60 seconds and the return code should be 0<br>2. Active firmware version and inactive firmware version remain unchanged<br>3. Committed image bank is updated to active image bank<br>4. No link flap is seen<br>5. Critical process such as `xcvrd`, `syncd`  `orchagent` does not ...
+| 4 | Firmware download validation with invalid firmware binary | Download an invalid firmware binary (any file not released by the vendor) | 1. The active firmware version does not change<br>2. The inactive firmware version remains unchanged or is set to `0.0.0` or `N/A`<br> 3.  No change in "Committed Image"<br>4. No link flap should be seen<br>5. The kernel has no error messages in syslog if ` ...
+| 5 | Firmware download interruption | 1. Start the firmware download and interrupt at the percentages specified by `firmware_download_interrupt_percentage`<br>2. Use the method specified in `firmware_download_interrupt_method` to interrupt the process:<br>"ctrl_c": Use CTRL+C or kill the download process<br> "sfputil_reset": reset the optics using sfputil reset<br>"optic_reinsert": remove the op ...
```
@az-pz Can you also ensure there is no link flap and there are no kernel errors seen (based on the `monitor_kernel_errors` attribute)?
I would recommend to create a separate section with the redundant expected results and just put a note to refer to that section.
You can do something similar to the below:
https://github.com/mihirpat1/sonic-mgmt/blob/ae573d7bd22020e9d112d9f59589efc158466f90/docs/testplan/transceiver/system_test_plan.md#standard-port-recovery-and-verification-procedure

mihirpat1 commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,302 @@
+| 3 | Firmware commit validation | 1. Execute firmware commit for an interface| 1. Firmware commit command should finish in 60 seconds and the return code should be 0<br>2. Active firmware version and inactive firmware version remain unchanged<br>3. Committed image bank is updated to active image bank<br>4. No link flap is seen<br>5. Critical process such as `xcvrd`, `syncd`  `orchagent` does not ...
+| 4 | Firmware download validation with invalid firmware binary | Download an invalid firmware binary (any file not released by the vendor) | 1. The active firmware version does not change<br>2. The inactive firmware version remains unchanged or is set to `0.0.0` or `N/A`<br> 3.  No change in "Committed Image"<br>4. No link flap should be seen<br>5. The kernel has no error messages in syslog if ` ...
+| 5 | Firmware download interruption | 1. Start the firmware download and interrupt at the percentages specified by `firmware_download_interrupt_percentage`<br>2. Use the method specified in `firmware_download_interrupt_method` to interrupt the process:<br>"ctrl_c": Use CTRL+C or kill the download process<br> "sfputil_reset": reset the optics using sfputil reset<br>"optic_reinsert": remove the op ...
+| 6 | CDB abort of firmware download | 1. Start the firmware download and interrupt using CDB abort command if `firmware_download_cdb_abort_support` flag is set | 1. Check that advertisement register for cdb abort is set if `firmware_download_cdb_abort_support` is set.<br> 2. Active firmware version remains unchanged<br>3. Inactive firmware version is invalid i.e. N/A or 0.0.0<br>4. No change in  ...
```
@az-pz Can you clarify that we need to interrupt the firmware download first and then issue the CDB abort command.
I would recommend making this as the first test case since CDB FW abort command is planned to be used for any firmware download process (irrespective of the previous state of the module).

az-pz committed:
Update docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md

Co-authored-by: mihirpat1 <112018033+mihirpat1@users.noreply.github.com>

az-pz committed:
Update test order and reduce redundant steps.

mssonicbld commented:
/azp run

az-pz committed:
Update test verification step.

az-pz committed:
Update formatting.

roy-sror commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,309 @@
+    <normalized_vendor_name_1>,<normalized_vendor_pn_1>,<firmware_version_1>,<firmware_binary_1>,<md5sum_1>
+    <normalized_vendor_name_1>,<normalized_vendor_pn_1>,<firmware_version_2>,<firmware_binary_2>,<md5sum_2>
+    <normalized_vendor_name_1>,<normalized_vendor_pn_1>,<firmware_version_3>,<firmware_binary_3>,<md5sum_3>
+    # Add more vendor part numbers as needed
```
I vote for JSON instead of CSV

roy-sror commented on docs/testplan/transceiver_onboarding/cmis_cdb_firmware_upgrade_testplan.md:
```diff
@@ -0,0 +1,309 @@
+
+**Core Normalization Rules:**
+
+1. **Character Replacement:**
```
why not regex? this is a standard way to represent strings in a dynamic format

roy-sror commented on docs/testplan/transceiver_onboarding/transceiver_onboarding_test_plan.md:
```diff
@@ -225,7 +537,8 @@ The following tests aim to validate the link status and stability of transceiver
 | Issue CLI command to shutdown a port | Validate link status using CLI configuration | Ensure that the link goes down |
 | Issue CLI command to startup a port | Validate link status using CLI configuration | Ensure that the link is up and the port appears in the LLDP table. |
-| In a loop, issue startup/shutdown command 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
+| In a loop, issue startup/shutdown command for a port 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
```
the number of iterations should be configurable by the user. the main concern with 100 or any other number is runtime

mihirpat1 commented:
@az-pz Please add the below TC as well
| Firmware version validation | 1. For transceivers with `gold_firmware_version` attribute, use `sfputil show fwversion <port>`.<br>2. If `dual_bank_supported` is true, verify both active and inactive firmware versions.<br>3. Compare with expected values from attributes. | Active and inactive firmware versions match corresponding values in attributes dictionary. |

mihirpat1 commented:
@az-pz as part of cleanup work on the EEPROM test plan, we removed the "Firmware version validation" test case since it is CMIS CDB-specific (uses `sfputil show fwversion` and validates active/inactive bank versions). As a result, the following three attributes are no longer referenced anywhere in the EEPROM test plan and should be moved to the CDB FW upgrade testplan's Attributes section instead:

| Attribute | Description |
|---|---|
| `dual_bank_supported` | Whether the transceiver supports dual-bank firmware |
| `gold_firmware_version` | Expected active/gold firmware version for CDB modules |
| `inactive_firmware_version` | Expected inactive bank firmware version for dual-bank CDB modules |

These are pure CDB FW upgrade concept and the upgrade testplan (this PR) is the correct home for them. Could you add them to the Attributes table in `cmis_cdb_firmware_upgrade_testplan.md`?

mihirpat1 commented:
MSFT ADO - 38346445
